prov/rxm/src/rxm_cq.c

```diff
@@ -353,12 +353,32 @@ static int rxm_lmt_handle_ack(struct rxm_rx_buf *rx_buf)
 	}
 }
 
+static inline size_t
+rxm_sar_calculate_recv_offset(struct rxm_ep *rxm_ep, size_t recv_seg_no)
+{
+	size_t offset = 0, i;
+
+	for (i = 0; ((i <= rxm_ep->sar_max_calc_seg_no) &&
+		     (i < recv_seg_no)); i++)
+		offset += RXM_SAR_DIVIDER << i;
+
+	if (recv_seg_no > (rxm_ep->sar_max_calc_seg_no + 1))
+		offset += (/* prev seg */(recv_seg_no - 1) -
+			   rxm_ep->sar_max_calc_seg_no) *
+			  rxm_ep->rxm_info->tx_attr->inject_size;
+
+	return offset;
+}
```

Comment:
Can we use ctrl_hdr.ctrl_data to to specify the offset? That may help us to avoid offset calculation on the receive side.

Reply:
OFI defines length of messages as size_t.
Yes, we can, but we may not fit into 62 bytes (2 bytes of SAR is used for flags) when SAR is used for large messages.

Reply:
But isn't ctrl_data field unused? Also the offset should fit well within 62 or 64 bytes.

Reply:
Yes, we can add an assert. Since `rxm_sar_ctrl_data->offset` is `uint32_t` we can add `assert(offset <= (uint32_t)-1)`

Reply:
Btw, the remove diffs are in the next patch. It should be moved here.

Reply:
Thanks! Done.
I meld the second commit with the first one, because they need to go together.